programs/ssl/ssl_base64_dump.c

```diff
@@ -0,0 +1,949 @@
+    {
+        CHECK_SSL_END( 4 );
+        uint32_t badmac_seen = ( (size_t) ssl[0] << 24 ) |
+                               ( (size_t) ssl[1] << 16 ) |
+                               ( (size_t) ssl[2] <<  8 ) |
+                               ( (size_t) ssl[3] );
+        ssl += 4;
+        printf( "\tbad MAC seen number                : %u\n", badmac_seen );
+
+        /* value 'in_window_top' from mbedtls_ssl_context */
+        printf( "\tlast validated record sequence no. : " );
+        CHECK_SSL_END( 8 );
+        print_hex( ssl, 8, 20, "" );
+        ssl += 8;
+
+        /* value 'in_window' from mbedtls_ssl_context */
+        printf( "\tbitmask for replay detection       : " );
+        CHECK_SSL_END( 8 );
+        print_hex( ssl, 8, 20, "" );
+        ssl += 8;
+    }
+
+    if( conf_dtls_proto )
+    {
+        CHECK_SSL_END( 1 );
+        printf( "\tDTLS datagram packing              : %s\n",
+                enable_disable( *ssl++ ) );
+    }
+
+    /* value 'cur_out_ctr' from mbedtls_ssl_context */
+    printf( "\toutgoing record sequence no.       : ");
+    CHECK_SSL_END( 8 );
+    print_hex( ssl, 8, 20, "" );
+    ssl += 8;
+
+    if( conf_dtls_proto )
+    {
+        CHECK_SSL_END( 2 );
+        uint16_t mtu = ( ssl[0] << 8 ) | ssl[1];
+        ssl += 2;
+        printf( "\tMTU                                : %u\n", mtu );
+    }
+
+
+    if( CONTEXT_CONFIG_ALPN_BIT & context_cfg_flag )
+    {
+        uint8_t alpn_len;
+
+        CHECK_SSL_END( 1 );
+        alpn_len = *ssl++;
+        printf_dbg( "ALPN length %u\n", (uint32_t) alpn_len );
+
+        printf( "\tALPN negotiation                   : " );
+        CHECK_SSL_END( alpn_len );
+        if( alpn_len > 0 )
+        {
+            if( strlen( (const char*) ssl ) == alpn_len )
+            {
+                printf( "%s\n", ssl );
+            }
+            else
+            {
+                printf( "\n" );
+                printf_err( "\tALPN negotiation is incorrect\n" );
+            }
+            ssl += alpn_len;
+        }
+        else
+        {
+            printf( "not selected\n" );
+        }
+    }
+
+    if( 0 != ( end - ssl ) )
+    {
+        printf_err( "%i bytes left to analyze from context\n", (int32_t)( end - ssl ) );
+    }
+    printf( "\n" );
+}
+
+int main( int argc, char *argv[] )
+{
+    enum { B64BUF_LEN = 4 * 1024 };
+    enum { SSLBUF_LEN = B64BUF_LEN * 3 / 4 + 1 };
+
+    uint8_t b64[ B64BUF_LEN ];
+    uint8_t ssl[ SSLBUF_LEN ];
+    size_t b64_counter = 0;
+
+    parse_arguments( argc, argv );
+
+    while( NULL != b64_file )
+    {
+        size_t ssl_len;
+        size_t b64_len = read_next_b64_code( b64, B64BUF_LEN );
+        if( b64_len > 0)
+        {
+            int ret;
+
+            printf( "\nDeserializing number %zu:\n",  ++b64_counter );
+
+            printf( "\nBase64 code:\n" );
+            print_b64( b64, b64_len );
+
+            ret = mbedtls_base64_decode( ssl, SSLBUF_LEN, &ssl_len, b64, b64_len );
```

Comment:
Before attempting to decode, I would recommend you check that the serialized data length is as expected and return an error if the base64 data is bigger than any valid data ever would be (in the biggest, fullest configuration), or smaller than any valid data ever would be (in the smallest configuration).

Reply:
I added checking length during reading base64 code from a file